### PR TITLE
Add IterativeSaver context manager and test

### PR DESCRIPTION
**Pull request by esoergel**
**DON'T MERGE** (yet), looking for feedback.

from https://github.com/dimagi/commcare-hq/pull/6163
@dannyroberts @czue 

This doesn't do anything to handle doc update conflicts.

One option for dealing with that would be to come up with a different pattern - one passing a transform function, so docs can be re-tried.  Something like:

```python
def remove_loc_types(domain_doc):
    if 'location_type' in domain_doc:  # don't bother needlessly saving docs
        domain_doc.pop('location_type')
        return domain_doc

iter_update(remove_loc_types, Domain.get_all_ids())
```

Alternatively, I could just catch any `ResourceConflict`s with this current version and log the docs that may have failed to update (any in that set), leaving it up to the user to follow up

This existing strategy without any error handling has worked in the past, and I don't wanna get too deep into worrying about potentially irrelevant edge cases before getting feedback.

### CODE REVIEW

**Comment by czue** [comment]
@esoergel I think you should use `FakeCouch` for the tests.

**Comment by czue** [comment]
(or real couch)

File: dimagi/utils/couch/bulk.py [hunk 1]
```diff
@@ -100,3 +100,33 @@ def wrapped_docs(cls, keys):
...
+        with IterativeSaver(db) as iter_db:
+            for doc in iter_docs(db)
+                iter_db.save(doc)
+    """
+    def __init__(self, database, chunksize=100):
+        self.db = database
+        self.chunksize = chunksize
+
+    def __enter__(self):
+        self.to_save = []
```
Comment (czue) [inline dimagi/utils/couch/bulk.py]: I don't know how fancy we want to get, but it could be nice to make `to_save` a dictionary, so if you try to save the same doc twice it's smart enough to only do it once

Comment (czue) [inline dimagi/utils/couch/bulk.py]: (in the same chunk)

**Comment by czue** [comment]
Not sure I would change a clean API to make error handling easier so not sure it makes sense to incorporate your example into this tool (though that seems a fine pattern to use).

I think ideal behavior would be:
- All docs that don't have conflicts are saved.
- All docs that have conflicts are raised as errors (maybe a custom exception type that returns a list of conflicted docs during commit?)

Not sure how hard it would be to produce something like that. You probably need to add some significant complexity to the commit method

File: dimagi/utils/couch/bulk.py [hunk 2]
```diff
@@ -100,3 +100,33 @@ def wrapped_docs(cls, keys):
...
+        return self
+
+    def commit(self):
+        self.db.bulk_save(self.to_save)
```
Comment (dannyroberts) [inline dimagi/utils/couch/bulk.py]: This will raise `BulkSaveError` if there's a resource conflict (even though all the docs that did _not_ confict will have gone through). If you catch the `BulkSaveError` as `e`, you can use `e.errors` to get the ids of all the docs that did not get saved. I would keep a running list of these and then raise your own `BulkSaveError` with those or have some other way for the caller to access that list.

Comment (esoergel) [inline dimagi/utils/couch/bulk.py]: Awesome, I *thought* there was something like that.  I'll for sure use that.

**Comment by esoergel** [comment]
Rewritten here: https://github.com/dimagi/commcare-hq/pull/6302
